syntax = "proto3";

package taxi_service;

// Servicios principales
service TaxiDatabaseService {
    rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
    rpc RegisterTaxi(RegisterTaxiRequest) returns (RegisterTaxiResponse);
    rpc UpdateTaxiPosition(UpdateTaxiPositionRequest) returns (UpdateTaxiPositionResponse);
    rpc CreateService(CreateServiceRequest) returns (CreateServiceResponse);
    rpc GetStatistics(GetStatisticsRequest) returns (GetStatisticsResponse);
    rpc GetAvailableTaxis(GetAvailableTaxisRequest) returns (stream Taxi);
}

// Mensajes base
message Position {
    double latitude = 1;
    double longitude = 2;
    string timestamp = 3;
}

message Taxi {
    string taxi_id = 1;
    string status = 2;
    Position current_position = 3;
    Position initial_position = 4;
    int32 total_services = 5;
    int32 successful_services = 6;
}

message Service {
    string service_id = 1;
    string taxi_id = 2;
    string status = 3;
    Position client_position = 4;
    Position taxi_initial_position = 5;
    string request_timestamp = 6;
    string completion_timestamp = 7;
}

// Health Check
message HealthCheckRequest {
    string component = 1;
}

message HealthCheckResponse {
    bool status = 1;
    string message = 2;
    string timestamp = 3;
}

// Registro de taxi
message RegisterTaxiRequest {
    string taxi_id = 1;
    Position initial_position = 2;
    string status = 3;
}

message RegisterTaxiResponse {
    bool success = 1;
    string message = 2;
    string taxi_id = 3;
}

// Actualización de posición
message UpdateTaxiPositionRequest {
    string taxi_id = 1;
    Position position = 2;
    string status = 3;
    string timestamp = 4;
}

message UpdateTaxiPositionResponse {
    bool success = 1;
    string message = 2;
    Position confirmed_position = 3;
}

// Creación de servicio
message CreateServiceRequest {
    string client_id = 1;
    Position client_position = 2;
    string timestamp = 3;
}

message CreateServiceResponse {
    bool success = 1;
    string service_id = 2;
    string assigned_taxi_id = 3;
    string message = 4;
    Position taxi_position = 5;
}

// Estadísticas
message GetStatisticsRequest {
    string time_range = 1;
}

message TaxiStats {
    string taxi_id = 1;
    int32 total_services = 2;
    int32 successful_services = 3;
    repeated Position historical_positions = 4;
    Position last_known_position = 5;
}

message GetStatisticsResponse {
    int32 total_services = 1;
    int32 completed_services = 2;
    int32 denied_services = 3;
    map<string, TaxiStats> taxi_statistics = 4;
    string timestamp = 5;
}

// Consulta de taxis disponibles
message GetAvailableTaxisRequest {
    Position reference_position = 1;
    double radius = 2;
}